samples/build-duration-logger/index.js

```diff
@@ -0,0 +1,255 @@
+            eventListeners: [
+                {
+                    eventName: 'Build',
+                    eventHandler:event => { this.enqueue(JSON.parse(event.data)); }
+                }
+            ],
+            retry: {
+                interval: 6000,
+                maxRetries: 30
+            }
+        });
+    }
+
+    enqueue(build) {
+        this.pendingBuilds.push(build);
+        this.processPendingBuilds();
+    }
+
+    processPendingBuilds() {
+        if (this.pendingBuilds.length > 0 && this.numBuildsInProcess < this.maxConcurrentBuildsToProcess) {
+            this.processBuild(this.pendingBuilds.shift());
+        }
+    }
+
+    createBuildStreamUrl(startTime) {
+        return `${this.gradleEnterpriseServerUrl}/build-export/v1/builds/since/${startTime}?stream`;
+    }
+
+    // Inspect the methods on the handler class to find any event handlers that start with 'on' followed by the event type like 'onBuildStarted'.
+    // Then take the part of the method name after the 'on' to get the event type.
+    getHandledEventTypesForHandlerClass(handlerClass) {
+        return Object.getOwnPropertyNames(handlerClass.prototype)
+            .filter(methodName => methodName.startsWith('on'))
+            .map(methodName => methodName.substring(2));
+    }
+
+    getAllHandledEventTypes() {
+        return new Set(this.eventHandlerClasses.reduce((eventTypes, handlerClass) => eventTypes.concat(this.getHandledEventTypesForHandlerClass(handlerClass)), []));
+    }
+
+    createBuildEventStreamUrl(buildId) {
+        const types = [...this.allHandledEventTypes].join(',');
+        return `${this.gradleEnterpriseServerUrl}/build-export/v1/build/${buildId}/events?eventTypes=${types}`;
+    }
+
+    // Creates a map of event type -> handler instance for each event type supported by one or more handlers.
+    createBuildEventHandlers(build) {
+        return this.eventHandlerClasses.reduce((handlers, handlerClass) => {
+            const addHandler = (type, handler) => handlers[type] ? handlers[type].push(handler) : handlers[type] = [handler];
+
+            const handler = new handlerClass.prototype.constructor(build);
+
+            this.getHandledEventTypesForHandlerClass(handlerClass).forEach(eventType => addHandler(eventType, handler));
+
+            if (Object.getOwnPropertyNames(handlerClass.prototype).includes('complete')) {
+                addHandler('complete', handler);
+            }
+
+            return handlers;
+        }, {});
+    }
+
+    processBuild(build) {
+        this.numBuildsInProcess++;
+        const buildEventHandlers = this.createBuildEventHandlers(build);
+        const buildEventStreamUrl = this.createBuildEventStreamUrl(build.buildId);
+
+        createServerSideEventStream(buildEventStreamUrl, {
+            oncomplete:  () => {
+                // there isn't an onclose event that we can listen to, but an error event does get triggered
```

Comment:
This comment needs an update as we now have the `oncomplete` method here and with this we know once the end of stream is reached. This comment is from where we used the `onerror` callback.

Shouldn't we call this as last  on the `oncomplete` handler such that all `complete()` methods get called first before we finish processing the builds?

Reply:
> Shouldn't we call this as last on the oncomplete handler such that all complete() methods get called first before we finish processing the builds?

I might not be understanding the comment but this is what is doing at the moment

Reply:
I have removed the comment (It was removed at some point but I've messed with Git and the comments somehow)